bin/varnishd/cache/cache_conn_pool.c

```diff
@@ -249,17 +255,35 @@ VCP_Rel(struct conn_pool **cpp)
 		(void)shutdown(pfd->fd, SHUT_RDWR);
 		cp->n_kill++;
 	}
-	while (cp->n_kill) {
-		Lck_Unlock(&cp->mtx);
-		(void)usleep(20000);
-		Lck_Lock(&cp->mtx);
-	}
+
+	Lck_Lock(&dead_pools_mtx);
+	VRBT_INSERT(vrb, &dead_pools, cp);
+	Lck_Unlock(&dead_pools_mtx);
 	Lck_Unlock(&cp->mtx);
```

Comment:
You should be able to unlock `cp->mtx` at the end of the foreach loop and keep `dead_pools_mtx` independent. At this point only waiters may have a reference of the pool and they are not concerned with connection pools cleanup, only individual connections.

We could even consider checking `n_kill` here to avoid a trip to the CLI hook:

```c
{
        /* ... */

        if (cp->n_kill == 0) {
                vcp_destroy(&cp);
                return;
        }

        Lck_Lock(&dead_pools_mtx);
        VRBT_INSERT(vrb, &dead_pools, cp);
        Lck_Unlock(&dead_pools_mtx);
}
```

Where `vcp_destroy()` would also be called from `VCP_RelPoll()`.

Reply:
Done :)